### PR TITLE
Atracsys support

### PR DESCRIPTION
**Pull request by markasselin**
Adding ability to use Atracsys tracker. This PR is mainly for review purposes until testing has been completed with the fusionTrack devices and any necessary changes made.

### CODE REVIEW

File: src/PlusDataCollection/Atracsys/AtracsysTracker.cxx [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+{
+  return this->GeometryId;
+}
+
+int AtracsysTracker::Marker::GetGeometryPrecsenceMask()
+{
+  return this->GeometryPresenceMask;
+}
+
+vtkSmartPointer<vtkMatrix4x4> AtracsysTracker::Marker::GetTransformToTracker()
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/AtracsysTracker.cxx]: The matrix cannot be deleted before returning, so use vtkMatrix4x4* instead of vtkSmartPointer.

File: src/PlusDataCollection/Atracsys/AtracsysTracker.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    int GetGeometryPrecsenceMask();
+    vtkSmartPointer<vtkMatrix4x4> GetTransformToTracker();
+    float GetFiducialRegistrationErrorMm();
+  private:
+    int Id; /*!< Tracking id */
+    int GeometryId;
+    vtkSmartPointer<vtkMatrix4x4> ToolToTracker;
+    int GeometryPresenceMask; /*!< Presence mask of fiducials expressed as
+                              * their geometrical indexes */
+    float FreMm; /*!< Registration mean ATRACSYS_ERROR (unit mm) */
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/AtracsysTracker.h]: rename to FiducialRegistrationErrorMm

File: src/PlusDataCollection/Atracsys/AtracsysTracker.h [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  class Marker
+  {
+  public:
+    Marker(int geometryId, vtkSmartPointer<vtkMatrix4x4> toolToTracker, int gpm, float freMm);
+    int GetGeometryID();
+    int GetGeometryPrecsenceMask();
+    vtkSmartPointer<vtkMatrix4x4> GetTransformToTracker();
+    float GetFiducialRegistrationErrorMm();
+  private:
+    int Id; /*!< Tracking id */
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/AtracsysTracker.h]: could you add a short comment explaining the difference between tracking and geometry id?

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+  AtracsysTracker::DEVICE_TYPE DeviceType = AtracsysTracker::UNKNOWN_DEVICE;
+
+  // matches plus tool id to .ini geometry file names/paths
+  std::map<std::string, std::string> PlusIdMappedToGeometryFilename;
+
+  // matches fusionTrack internal tool geometry ID to Plus tool ID for updating tools
+  std::map<int, std::string> FtkGeometryIdMappedToToolId;
+
+  // Atracsys API wrapper class handle
+  AtracsysTracker* Tracker = new AtracsysTracker();
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Why not simply AtracsysTracker Tracker?
If we do need to use new operator then it has to be deleted in the destructor.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+  if (this->Internal->MaxMissingFiducials < 0 || this->Internal->MaxMissingFiducials > 3)
+  {
+    LOG_WARNING("Invalid maximum number of missing fiducials provided. Must be between 0 and 3 inclusive. Maximum missing fiducials has been set to its default value of 0.");
+    this->Internal->MaxMissingFiducials = 0;
+  }
+
+  XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(float, MaxMeanRegistrationErrorMm, this->Internal->MaxMeanRegistrationErrorMm, deviceConfig);
+  if (this->Internal->MaxMeanRegistrationErrorMm < 0.1 || this->Internal->MaxMeanRegistrationErrorMm > 5.0)
+  {
+    LOG_WARNING("Invalid maximum mean registration error provided. Must be between 0.1 and 5 mm inclusive. Maximum mean registration error has been set to its default value of 5.0mm.");
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Warning message states that it is set to 5.0mm, while we actually set it to 2.0mm.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    this->Internal->MaxMeanRegistrationErrorMm = 2.0;
+  }
+
+  XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(int, ActiveMarkerPairingTimeSec, this->Internal->ActiveMarkerPairingTimeSec, deviceConfig);
+  if (this->Internal->ActiveMarkerPairingTimeSec > 60)
+  {
+    LOG_WARNING("Are you sure you want to leave active marker pairing enabled for more than 60 seconds?");
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: This should not be a question - the user has no way of answer it. Rephrase it to a simple statement, something like: "Warning: active marker pairing time is set to ...sec. Tracking will not start until this time period is over."

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        return PLUS_FAIL;
+      }
+    }
+    else if (toolTrackingType == PASSIVE)
+    {
+      // passive tool, load into PlusIdMappedToGeometryFilename to have geometry loaded in InternalConnect
+      const char* geometryFile;
+      if ((geometryFile = toolDataElement->GetAttribute("GeometryFile")) != NULL)
+      {
+        std::pair<std::string, std::string> thisTool(toolId, geometryFile);
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: why don't we use a map<string,string> instead?

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+
+  return PLUS_SUCCESS;
+}
+
+//----------------------------------------------------------------------------
+PlusStatus vtkPlusAtracsysTracker::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  LOG_TRACE("vtkPlusAtracsysTracker::WriteConfiguration");
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(deviceConfig, rootConfigElement);
+  return PLUS_SUCCESS;
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Write out all elements that may be changed using Set...() methods

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+  LOG_TRACE("vtkPlusAtracsysTracker::Probe");
+  return PLUS_SUCCESS;
+}
+
+//----------------------------------------------------------------------------
+PlusStatus vtkPlusAtracsysTracker::InternalConnect()
+{
+  LOG_TRACE("vtkPlusAtracsysTracker::InternalConnect");
+
+  ATRACSYS_RESULT result;
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: merge with next line

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+  else if (result == AtracsysTracker::ATRACSYS_RESULT::WARNING_CONNECTED_IN_USB2)
+  {
+    LOG_WARNING(this->Internal->Tracker->ResultToString(result));
+  }
+
+  // get device type
+  this->Internal->Tracker->GetDeviceType(this->Internal->DeviceType);
+
+  // if spryTrack, setup for onboard processing and disable extraneous marker info streaming
+  if (this->Internal->DeviceType == AtracsysTracker::DEVICE_TYPE::SPRYTRACK_180)
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Could this low-level configuration pushed down into the AtracsysTracker class?
It would be nice if users of AtracsysTracker class would not need to know model-specific settings.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+  std::map<std::string, std::string>::iterator it;
+  for (it = begin(this->Internal->PlusIdMappedToGeometryFilename); it != end(this->Internal->PlusIdMappedToGeometryFilename); it++)
+  {
+    // load user defined geometry file
+    // TODO: add check for conflicting marker IDs
+    std::string geomFilePath = vtkPlusConfig::GetInstance()->GetDeviceSetConfigurationPath(it->second);
+    int geometryId;
+    if ((result = this->Internal->Tracker->LoadMarkerGeometry(geomFilePath, geometryId)) != ATR_SUCCESS)
+    {
+      LOG_ERROR(this->Internal->Tracker->ResultToString(result));
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Also log the geomFilePath that failed to load.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+
+  // pair active markers
+  if ((result = this->Internal->Tracker->EnableWirelessMarkerPairing(true)) != ATR_SUCCESS)
+  {
+    LOG_ERROR(this->Internal->Tracker->ResultToString(result));
+    return PLUS_FAIL;
+  }
+  LOG_INFO(endl << " *** Put active markers in front of the tracker to pair ***" << endl);
+
+  // sleep while waiting for tracker to pair active markers
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: use vtkPlusAccurateTimer::Delay instead

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+  // make LED blue during pairing
+  this->Internal->Tracker->SetUserLEDState(0, 0, 255, 0);
+
+  // pair active markers
+  if ((result = this->Internal->Tracker->EnableWirelessMarkerPairing(true)) != ATR_SUCCESS)
+  {
+    LOG_ERROR(this->Internal->Tracker->ResultToString(result));
+    return PLUS_FAIL;
+  }
+  LOG_INFO(endl << " *** Put active markers in front of the tracker to pair ***" << endl);
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Log messages are not really for displaying instructions to users. Change it to LOG_INFO("Active marker pairing period started"). Do the same for end of marker pairing period. Instructions for users can be implemented at the application level or described in user manual.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+    return PLUS_FAIL;
+  }
+
+  // make LED green, pairing is complete
+  this->Internal->Tracker->SetUserLEDState(0, 255, 0, 0);
+
+  // TODO: check number of active markers paired
+  // TODO: print info about paired markers here
+  
+  cout << "Additional info about paired markers:" << this->Internal->Tracker->GetMarkerInfo() << std::endl;
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Device classes should not print anything on cout.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+
+  std::vector<AtracsysTracker::Marker> markers;
+
+  ATRACSYS_RESULT result = result = this->Internal->Tracker->GetMarkersInFrame(markers);
+  if (result != ATR_SUCCESS && result != AtracsysTracker::ATRACSYS_RESULT::ERROR_NO_FRAME_AVAILABLE)
+  {
+    LOG_ERROR(this->Internal->Tracker->ResultToString(result));
+    return PLUS_FAIL;
+  }
+  else if (result == AtracsysTracker::ATRACSYS_RESULT::ERROR_NO_FRAME_AVAILABLE)
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: move this before the first check and then you can simplify

result != ATR_SUCCESS && result != AtracsysTracker::ATRACSYS_RESULT::ERROR_NO_FRAME_AVAILABLE

to

result != ATR_SUCCESS

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+      toolUpdated = true;
+      PlusTransformName toolTransformName(it->second, this->GetToolReferenceFrameName());
+      std::string toolSourceId = toolTransformName.GetTransformName();
+      ToolTimeStampedUpdate(toolSourceId, mit->GetTransformToTracker(), TOOL_OK, this->FrameNumber, unfilteredTimestamp);
+    }
+
+    if (!toolUpdated)
+    {
+      // tool is not seen in this frame
+      vtkSmartPointer<vtkMatrix4x4> emptyTransform = vtkSmartPointer<vtkMatrix4x4>::New();
```
Comment (lassoan) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: You can replace

    vtkSmartPointer<vtkMatrix4x4> emptyTransform = vtkSmartPointer<vtkMatrix4x4>::New();
    emptyTransform->Identity(); 

by

    vtkNew<vtkMatrix4x4> emptyTransform;

File: src/PlusDataCollection/CMakeLists.txt [hunk 17]
```diff
@@ -226,14 +226,7 @@ IF(PLUS_USE_OPTITRACK)
...
     )
-
-  IF (OPTITRACK_MSVC80_OPENMP_DIR)
-    LIST(APPEND External_Libraries_Install
-      ${OPTITRACK_MSVC80_OPENMP_DIR}/Microsoft.VC80.OpenMP.manifest
-      ${OPTITRACK_MSVC80_OPENMP_DIR}/vcomp.dll
-      )
-  ENDIF()
-
+
```
Comment (lassoan) [inline src/PlusDataCollection/CMakeLists.txt]: Was this change intentional?

File: src/PlusDataCollection/CMakeLists.txt [hunk 18]
```diff
@@ -589,6 +582,71 @@ IF(PLUS_USE_OPTICAL_MARKER_TRACKER)
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/Atracsys
+    ${ATRACSYS_SDK_INCLUDE_DIR}
+    )
+
+  LIST(APPEND ${PROJECT_NAME}_LIBS
+    ${ATRACSYS_SDK_LIBRARY}
+    )
+  
+  IF(WIN32)
+    IF(CMAKE_CL_64)
```
Comment (lassoan) [inline src/PlusDataCollection/CMakeLists.txt]: use void p size - https://cmake.org/cmake/help/v3.11/variable/CMAKE_CL_64.html

File: src/PlusDataCollection/Atracsys/AtracsysTracker.cxx [hunk 19]
```diff
@@ -0,0 +1,799 @@
+/*=Plus=header=begin======================================================
+Program: Plus
+Copyright (c) Laboratory for Percutaneous Surgery. All rights reserved.
+See License.txt for details.
+=========================================================Plus=header=end*/
+
+#include "AtracsysTracker.h"
+
+// System includes
+#include <string>
```
[the remark this answers is not in the document]
Comment (markasselin) [reply src/PlusDataCollection/Atracsys/AtracsysTracker.cxx]: Within the groups, or as one big long list of includes?

[the remark this answers is not in the document]
Comment (lassoan) [reply src/PlusDataCollection/Atracsys/AtracsysTracker.cxx]: Sort alphabetically within a group.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx (hunk 7 above)
Comment (markasselin) [reply src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: We do, I just create the std::pair to make adding to the map more intuitive

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx (hunk 8 above)
Comment (markasselin) [reply src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Not sure what you mean here - could you point me to an example please?

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx (hunk 7 above)
Comment (lassoan) [reply src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: std::pair is probably the most useless container. Since "first" and "second" members are not named, they don't make things very clear or self-explaining.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx (hunk 8 above)
Comment (lassoan) [reply src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: In general, attributes that you read from XML should be written, too, especially if there is public API to change them.

**Comment by markasselin** [comment]
Ready to merge
